### ISSUE TITLE
bug in pnet for receiving UDP packets?

### ISSUE THREAD

**Issue by stglasauer** [Psychtoolbox-3/Psychtoolbox-3#699]
Using pnet to read UDP packets works for `uint8` below 128, but not above. Sending 128 will result in receiving 255.
If you send `int8`, then the negative numbers don't work (e.g. sending anything below -1 will result in receiving -1).
This happens in general, the highest bit of every byte seems to be treated somehow differently, if it is 1.

**To Reproduce**
1) write a `uint8` to a local UDP port (I used the `udp` function from the Instrument Control Toolbox)
2) read from the same port, either with another `udp` instance or with `udpin=pnet('udpsocket',p);`

here is a complete test script (run as is to test :
```
p=7050; % port
usepnet=1; % set to 0 to use udp for receiving
precision='uint8'; % alternatively use 'single'

if usepnet
    udpin=pnet('udpsocket',p);
    pnet(udpin,'setreadtimeout',1.0);
else
    udpin = udp();
    udpin.LocalPort=p;
    udpin.TimeOut=1;
    fopen(udpin);
end

udpout = udp('127.0.0.1',p);
fopen(udpout);

c=126:129;

for i=1:numel(c)
if strcmp(precision,'single')
    d=single(c(i));
else
    d=uint8(c(i));
end

disp(['writing ' num2str(d) ' to port ' int2str(p) ' ' dec2bin(d)])
fwrite(udpout,d,precision)

pause(0.1)

if usepnet
    len=pnet(udpin,'readpacket');
    A=pnet(udpin,'read');
else
    [A,len,msg]=fread(udpin);
end

if len>0
    r=swapbytes(typecast(uint8(A)',precision));
    disp(['received ' num2str(r) ' ' dec2bin(r)])
else
    disp('error, nothing received')
end

end
if usepnet
    pnet(udpin,'close');
    pnet('closeall');
else
    fclose(udpin)
end

fclose(udpout)
instrreset
```

Output with pnet:
writing 129 to port 7050 10000001
received 255 11111111

Output with udp:
writing 129 to port 7050 10000001
received 129 10000001

**Expected behavior**
same behavior as `udp`, and receiving what has been sent.

**Desktop (please complete the following information):**
 - OS: macOS 10.15.7 and Windows-10
 - Runtime: Matlab R2020a
 - Hardware [e.g., AMD Radeon Pro W4100, Onboard sound chip]
 - Version Psychtoolbox 3.0.16

**Additional context**
There was a comment in 2016 on Matlab File Exchange for pnet about this (even though the behavior seems different now):
_"All incoming UDP data is being truncated at 7 bits. ... Looking at pnet.c I seem some double casting (char)(unsigned char) which looks a little suspicious."_

I hope there's a way to fix this, since using pnet to receive UDP is really helpful!

**Comment by kleinerm** [comment]
Can you get your lab to buy some priority support (our way of letting you contribute some money to the maintenance and development of PTB)? Once that is done, i can handle this as a support incident with some priority.

See this link for the procedure: http://psychtoolbox.org/prioritysupport.html

Thanks and happy new year
-mario

**Comment by stglasauer** [comment]
Hi Mario,

Already done, bought it yesterday (after filing the report) …  invoice attached!

It’s a great idea that you guys offer priority support, I’m happy to support you with this!

Thanks, and Happy New Year!

Stefan

**Comment by kleinerm** [comment]
Hi Stefan,
happy new year!

The invoice wasn't attached, but also the invoice isn't what i need. In fact, i prefer to **not see that at all**, to simplify customer data protection and also to prevent somebody else abusing your purchase.

What i do need for my accounting here is an authorization token, as generated from the order id and license key on that invoice, with our little helper script PsychPaidSupportAndServices(). See step 2 and onward on:

http://psychtoolbox.org/prioritysupport.html

Once i got that, i can look into it. Probably next week. Thanks for your financial support!

The community membership + priority support model is supposed to be our primary source of funding in the long term. The idea is that many/most labs which can afford it (and hopefully at this price point it is very affordable to most labs) will buy it, _even without specific need for priority support_. Basically like a fire insurance or subscription to a gym or club one never uses :). That's the only elegant way of funding general PTB development and maintenance that i find aligns best with both our interests and those of our users and of open science in general. This is also the outcome of a user survey we did between 2016-2018, so one would hope that labs are onboard with things they proposed and agreed to. So far the uptake after 1.5 months is very modest, at this level it wouldn't work in a meaningful way. So spread the word among your colleagues if possible, it is very neccessary!

**Comment by stglasauer** [comment]
Hi Mario,
sorry, I didn't understand that I had to get this token ... probably overlooked it, sorry.
But here's the token:
XYMZ6PUQ-202111523652:e3e268a6c5918a97810ef95a390bbaa474a446563660f39fd9d76e0d874eb375
Hope that works now!

**Comment by kleinerm** [comment]
Ok, that's what i needed. After checking the code and help text of pnet, in the end it turns out this is an omission in your code:

If you replace ``A=pnet(udpin,'read');`` by ``A=pnet(udpin,'read', 1, precision);`` then it works.

Then you also don't need those typcast statement. A is already of class uint8.

help pnet for the 'Read' subfunction has this snippet:

Syntax:

data=pnet(con,'read' [,size] [,datatype] [,swapping] [,'view'] [,'noblock'])

...The **datatype** option is a string containing the name of the
      datatype, if **not specified its by default 'char'**. Supported datatypes is
      the same as for the 'write' command. ...

To if the optional 'datatype' is left out, it defaults to interpret and return all data as Matlab char()'s, and that ends up with some ASCII encoding in range 0 - 127 afaics. The 'uint8' or 'int8' types instead allow for 0 to 255, or -128 to 127.

Does this solve the problem for you?

-mario

[This used up 35 minutes of work time out of the 60 minutes available to you.]

**Comment by stglasauer** [comment]
Oh ... ok, shame on me, I thought that it's a pnet problem, because it does not happen when using Matlab's udp function. And that's why I didn't realize that pnet caps the first bit for char() and that setting the datatype would solve this problem ... and yes, it solves the problem. 

Thanks! Case closed!

Stefan

**Comment by kleinerm** [comment]
Just for clarification: pnet itself doesn't cap the bit, but it returns data in the Matlab mxChar data type by default. Matlab caps the uppermost bit, apparently treating such characters as 7-bit ASCII. Octave doesn't, as their string/character handling differs a bit, so there a char would have gone through with your code.

**Comment by stglasauer** [comment]
Thanks for clarifying this!
